### ISSUE TITLE
Execute common code for all Notebooks

### ISSUE THREAD

**Issue by borgdrone7** [nteract/papermill#550]
Hi,
I have a use case where we have a platform in which users will write Notebooks for some analytics processing. However there is some common code needed for all of the Notebooks at start (import, initialization, etc.) and some cleanup code at the end. So currently that same code is copy pasted in all the Notebooks and only middle part which users write is changed for each Notebook.
Is it possible to execute some code before execute_notebook call so that Notebook part is run in the same context, and then after execute_notebook call some other code. Basically I am asking if it is possible to have execute_notebook execute in python current process context (including all imported modules, etc.)? If that's not possible is it at least possible to dynamically add Notebook nodes at start/end of notebook before executing? Even if it is possible to execute 3 notebooks in the same context (init, script, end) that would also be good enough.

Thanks

**Comment by rgbkrk** [comment]
One of the core tenents in the project is that authoring and opening a notebook later should be a reproducible and re-runnable notebook. If you have common code at the start, you don't want a new cell to be re-injected every time you run it again. At the same time, authors should be running their notebook from start-to-finish without relying on pipelines to add additional code.

My recommendation is to wrap this common code in a library with a simple import + call `import borgdrone; borgdrone.init()`

**Comment by borgdrone7** [comment]
Hi,
Thank you for the quick reply. 
I understand that Notebooks in general should be self contained and that in most cases makes perfect sense, however in our case we intentionally want these notebooks to be part of the pipeline. They make no sense to be executed on their own as they will receive dynamic data and lots of other stuff needs to happen before/after the Notebook execution part of the pipeline. Basically users will write analytics transformation of the data and once they execute pipeline (outside of Jupyter, from custom GUI which includes scheduling and other stuff) which will go through lots of stages then execute all the Notebooks users can add as middle-ware and after that common code will be run which will as a result make them able to see the graphs of their data transformations when applied in Grafana. Making these testable and self contained is impractical as the result can only be represented graphically to make sense and must be based on lots of data (which users are already familiar with and can easily check the results in graphs). 
In short, these notebooks are intended to be part of bigger pipeline and, at least in our case, having 2 lines of code for each Notebook, one before and one after the code will probably be a small distraction for the users. They cannot do any meaningful change in those lines and were not meant to. 
Anyway, Papermill is a life saver for us and even for pipelines scenario I think it is really good. I just think that it might make sense to be able to re-inject new cell each time the Notebook is run in some cases, like we have. But if it is against core tenets of the project I don't expect it to be implemented and probably most people can live without it.
